### PR TITLE
Add OpenCL support

### PR DESCRIPTION
**Pull request by glupescu**
This set of commits introduces single GPU OpenCL experimental support for solano-perf-libs. The CUDA portions should be unaltered, as the source code is not shared. Only single GPU with 1 context and 1 queue is supported at this point.

OpenCL frameworks entails a separation of HOST and DEVICE code, which complicates things, alongside the manual search and setup of platforms of devices. These are done automatically in CUDA, but must be handled manually in OpenCL.

Most *.cu files have been transformed in the following:
-> HOST / CPU code result in *.cpp files, compiled by chosen compiler (e.g. g++)
-> DEVICE / {GPU, CPU, FPGA} code result in *.h files, included in the binary and compiled at runtime given the first found device of chosen type (currently CL_DEVICE_TYPE_GPU). All these files are in opencl-platform/ directory.

Every function which runs on the DEVICE, has a cl_check_init which if OpenCL has not been init (device search/setup, kernel comilation etc), it will init. The init process takes some time (typically 1-5 sec), but is performed only once.

Various results of the OpenCL port, cl_ed25519_verify

#### CUDA initial code running on Tesla K40m

./debug/cuda_ed25519_verify 100000 1 1 1 1 1
time diff: 1188735.000000 total: 100000 sigs/sec: 84123.038356

#### OpenCL ported code running on Tesla K40m

./debug/cl_ed25519_verify 100000 1 1 1 1 1
OpenCL platform query & init...
OpenCL init devices query type: GPU
Platforms found: 2
Platform 0 Intel(R) Corporation OpenCL 1.2 LINUX
Platform 1 NVIDIA Corporation OpenCL 1.2 CUDA 9.1.84
        Devices found 3
        Device 0 Tesla K40m
        Device 1 Tesla K40m
        Device 2 Tesla K40m
Compiling aes-cbc kernels
Compiling chacha_cbc kernels
Compiling verify kernels
time diff: 336381.000000 total: 100000 sigs/sec: 297281.951121

#### OpenCL ported code running on Intel Xeon E5-2670

./debug/cl_ed25519_verify 100000 1 1 1 1 1
OpenCL platform query & init...
OpenCL init devices query type: CPU
Platforms found: 2
Platform 0 Intel(R) Corporation OpenCL 1.2 LINUX
        Devices found 1
        Device 0       Intel(R) Xeon(R) CPU E5-2670 v2 @ 2.50GHz
Platform 1 NVIDIA Corporation OpenCL 1.2 CUDA 9.1.84
Compiling aes-cbc kernels
Compiling chacha_cbc kernels
Compiling verify kernels
8 warnings generated.
time diff: 1578123.000000 total: 100000 sigs/sec: 63366.416940

### CODE REVIEW

**Comment by glupescu** [comment]
Though this commit should not alter in any way the current behavior I've pushed it to start also a discussion on how the OpenCL implementation should be.

OpenCL has the benefit to support many types of architectures (CPU, GPU, FPGA) from many vendors (Intel, Amd, Nvidia, Samsung etc), but at the same time introduces the following issues:

1. Significant code complexity, because host and device have different source codes, and the compilation is done at runtime (practical purposes on portability).
For a production release the kernel source code should be embedded.

2. Platform and device selection which is not automatic, but is queried each time and at the developer's discretion

**Comment by sakridge** [comment]
Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519

I would say including the CL source as static string in the binary is fine.

File: src/cuda-headers/gpu_common.h [hunk 1]
```diff
@@ -21,4 +33,61 @@ inline void cuda_assert(cudaError_t err, const char *file, int line)
     }
 }
 
+/************************************
```
Comment (sakridge) [inline src/cuda-headers/gpu_common.h]: I would split this into two files something like: gpu_cuda_common.h and gpu_opencl_common.h. The OpenCL and cuda will not share this logic probably.

**Comment by glupescu** [comment]
> Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
> https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519
> 
> I would say including the CL source as static string in the binary is fine.

Aiming first to have all the pieces converted and then will look into correctness. After correctness I will look into performance.

Do you have any idea of the candidate typical architecture ? Are we expecting a high end GPU or any GPU including integrated ones ? This will dictate the approach to have on optimization.

**Comment by sakridge** [comment]
> > Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
> > https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519
> > I would say including the CL source as static string in the binary is fine.
> 
> Aiming first to have all the pieces converted and then will look into correctness. After correctness I will look into performance.
> 
> Do you have any idea of the candidate typical architecture ? Are we expecting a high end GPU or any GPU including integrated ones ? This will dictate the approach to have on optimization.

I would say leaning towards higher end GPUs. I wouldn't spend much time optimizing for integrated ones.

**Comment by glupescu** [comment]
> Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
> https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519
> 
> I would say including the CL source as static string in the binary is fine.

Can you provide some sample scenarios ? (args for running cuda-ecc-ed25519)

**Comment by glupescu** [comment]
> Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
> https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519
> 
> I would say including the CL source as static string in the binary is fine.

This is the path I went with. Currently the OpenCL kernel files can be found in src\opencl-platform and are included at compile time in the binary. They are however compiled at runtime after device selection. Currently they compile both on Linux/CUDA and Windows/AMDAPP. The compilation is done only once, upon device search & init.

**Comment by sakridge** [comment]
> > Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
> > https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519
> > I would say including the CL source as static string in the binary is fine.
> 
> Can you provide some sample scenarios ? (args for running cuda-ecc-ed25519)

Here's one example with 100k signatures:
```
sakridge@valverde:~/src/solana-perf-libs/src/release$ ./cuda_ed25519_verify 100000 1 1 4 10 1
time diff: 922415.000000 total: 4000000 sigs/sec: 4336442.924280
```

**Comment by sakridge** [comment]
> > Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
> > https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519
> > I would say including the CL source as static string in the binary is fine.
> 
> This is the path I went with. Currently the OpenCL kernel files can be found in src\opencl-platform and are included at compile time in the binary. They are however compiled at runtime after device selection. Currently they compile both on Linux/CUDA and Windows/AMDAPP. The compilation is done only once, upon device search & init.

Ok cool. We need to have some kind of runtime selection for opencl/cuda.

**Comment by glupescu** [comment]
> > > Hi @glupescu great effort. I would say the verify functionality is the most important however. If you can focus on this part of the code:
> > > https://github.com/solana-labs/solana-perf-libs/tree/master/src/cuda-ecc-ed25519
> > > I would say including the CL source as static string in the binary is fine.
> > 
> > 
> > This is the path I went with. Currently the OpenCL kernel files can be found in src\opencl-platform and are included at compile time in the binary. They are however compiled at runtime after device selection. Currently they compile both on Linux/CUDA and Windows/AMDAPP. The compilation is done only once, upon device search & init.
> 
> Ok cool. We need to have some kind of runtime selection for opencl/cuda.

Currently libraries generated are different so the compilation of the binary (which libraries it links with) would determine whether CUDA or OpenCL is used.

File: src/cuda-headers/gpu_common.h [hunk 2]
```diff
@@ -21,4 +21,4 @@ inline void cuda_assert(cudaError_t err, const char *file, int line)
     }
 }
```
Comment (sakridge) [inline src/cuda-headers/gpu_common.h]: can you revert?

File: src/opencl-crypt/aes.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2002-2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
```
Comment (sakridge) [inline src/opencl-crypt/aes.h]: Can we remove everything but the verify for now?

Comment (glupescu) [reply src/opencl-crypt/aes.h]: So have only the files needed to generate and run cl_ed25519_verify (OpenCL variant of ed25519_verify) ?

File: src/cuda-headers/gpu_common.h (hunk 2 above)
Comment (glupescu) [reply src/cuda-headers/gpu_common.h]: will do

**Pull request closed without merging**